observation_portal/proposals/viewsets.py

```diff
@@ -34,3 +78,130 @@ class SemesterViewSet(viewsets.ReadOnlyModelViewSet):
+            active=True, non_science=False
+        ).prefetch_related(
+            'sca', 'membership_set', 'membership_set__user', 'membership_set__user__profile',
+            'semester_set', 'timeallocation_set'
+        ).distinct().order_by('sca__name')
+        results = []
+        for proposal in proposals:
+            results.append({
+                'id': proposal.id,
+                'title': proposal.title,
+                'abstract': proposal.abstract,
+                'allocation': proposal.allocation(semester=semester),
+                'pis': [
+                    {
+                        'first_name': mem.user.first_name,
+                        'last_name': mem.user.last_name,
+                        'institution': mem.user.profile.institution
+                    } for mem in proposal.membership_set.all() if mem.role == Membership.PI
+                ],
+                'sca_id': proposal.sca.id,
+                'sca_name': proposal.sca.name,
+                'semesters': proposal.semester_set.distinct().values_list('id', flat=True)
+            })
+        return Response(results)
+
+    @action(detail=True, methods=['get'], permission_classes=(IsAdminUser,))
+    def timeallocations(self, request, pk=None):
+        timeallocations = self.get_object().timeallocation_set.prefetch_related(
+            'proposal', 'proposal__membership_set', 'proposal__membership_set__user'
+        ).distinct()
+        results = []
+        for timeallocation in timeallocations:
+            memberships = timeallocation.proposal.membership_set
+            timeallocation_dict = timeallocation.as_dict(exclude=['proposal', 'semester'])
+            timeallocation_dict['proposal'] = {
+                'notes': timeallocation.proposal.notes,
+                'id': timeallocation.proposal.id,
+                'tac_priority': timeallocation.proposal.tac_priority,
+                'num_users': memberships.count(),
+                'pis': [
+                    {
+                        'first_name': mem.user.first_name,
+                        'last_name': mem.user.last_name
+                    } for mem in memberships.all() if mem.role == Membership.PI
+                ]
+            }
+            results.append(timeallocation_dict)
+        return Response(results)
+
+
+class MembershipViewSet(ListAsDictMixin, DetailAsDictMixin, mixins.DestroyModelMixin, viewsets.ReadOnlyModelViewSet):
+    http_method_names = ('get', 'head', 'options', 'post', 'delete')
+    filter_backends = (DjangoFilterBackend, filters.OrderingFilter,)
+    filter_class = MembershipFilter
+    serializer_class = MembershipSerializer
+
+    def get_queryset(self):
+        if self.request.user.is_staff and self.request.user.profile.staff_view:
+            return Membership.objects.all()
+        else:
+            users_memberships = self.request.user.membership_set.all()
+            pi_memberships_of_users_proposals = Membership.objects.filter(
```

Comment:
this line is getting the memberships for the PIs for proposals you are a part of right? What is this used for?

Reply:
That's right. I just wrote this queryset to reflect the memberships that a user should be allowed to see, based on what we decided was acceptable for GDPR. This includes the details of the PIs on their proposals. So, a user should be allowed to see:

- all of their own memberships
- the members that are PIs on their proposals
- all of the members of any proposals on which they are the PI

Do you think that's overkill?

Reply:
I guess I don't know what view its shown in, so how the info is useful for a user... If that's what we decided though its fine, I was just wondering how its used since it didn't seem useful.